### ISSUE TITLE
MESH ASSOCIATION field-line gated at v751, should be v740sp05 (matches BEGIN/END OF MESH)

### ISSUE THREAD

**Issue by rubber-duck-53** [abaplint/abaplint#4105]
## Summary

> _Disclaimer: This issue report was drafted with AI assistance. If something looks odd or you need more details, please let me know._
>
`type_mesh.ts` gates the mesh-node-with-`ASSOCIATION` statement at `Release.v751`, but the enclosing `type_mesh_begin.ts` and `type_mesh_end.ts` are gated at `Release.v740sp05`. This is internally inconsistent (a `BEGIN OF MESH` / `END OF MESH` block is only meaningful with association components between it), and the SAP-documented introduction release for the full mesh syntax — including the inline `ASSOCIATION ... TO ... ON` clauses — is 7.40 SP05, not 7.51.

Consequence: with `syntax.version.release = "v750"`, abaplint's `parser_error` rule flags every `ASSOCIATION` field-line inside a `MESH` block, even though the code activates cleanly on a real SAP_BASIS 7.50 system.

## Environment

- abaplint: `2.119.56`
- Node: 18+ (default)
- OS: Windows 11
- Config: `syntax.version.release = "v750"`, `language = "Normal"`
- Rule enabled: `parser_error`

## Reproduction

`abaplint.json` (minimal):

```json
{
  "global": {
    "files": "/src/**/*.*"
  },
  "syntax": {
    "version": "v750",
    "errorNamespace": ".*"
  },
  "rules": {
    "parser_error": true
  }
}
```

`src/zcl_mesh_demo.abap` (minimal — types only, so package layout is not material):

```abap
TYPES BEGIN OF ty_parent.
TYPES   id   TYPE string.
TYPES   name TYPE string.
TYPES END OF ty_parent.
TYPES tt_parent TYPE STANDARD TABLE OF ty_parent WITH DEFAULT KEY.

TYPES BEGIN OF ty_child.
TYPES   id        TYPE string.
TYPES   parent_id TYPE string.
TYPES END OF ty_child.
TYPES tt_child TYPE STANDARD TABLE OF ty_child WITH DEFAULT KEY.

TYPES BEGIN OF ty_detail.
TYPES   ref_id TYPE string.
TYPES   value  TYPE string.
TYPES END OF ty_detail.
TYPES tt_detail TYPE STANDARD TABLE OF ty_detail WITH DEFAULT KEY.

TYPES BEGIN OF MESH ty_mesh.
TYPES parents  TYPE tt_parent.
TYPES children TYPE tt_child
      ASSOCIATION to_parent
          TO parents
          ON id = parent_id.
TYPES details  TYPE tt_detail
      ASSOCIATION to_child
          TO children
          ON id = ref_id.
TYPES END OF MESH ty_mesh.
```

### Expected

No parser errors. The syntax is documented in the SAP ABAP Keyword Documentation for release 7.40 (SP05 onwards) and is unchanged in 7.50.

### Actual

`parser_error` findings on each `TYPES: <node> TYPE ... ASSOCIATION ...` line inside the `MESH` block.

## Evidence that the correct floor is `v740sp05`, not `v751`

1. **SAP 7.40 keyword documentation (static, server-rendered)** — the SP05 news page "Meshes in Release 7.40, SP05" introduces meshes with the full syntax:

   > `TYPES BEGIN OF MESH ... TYPES snode ... ASSOCIATION to_node TO tnode ON tcomp1 = scomp1 [AND ...]. ... TYPES END OF MESH`

   as a single new `TYPES` variant. The `ASSOCIATION ... TO ... ON` field-line is documented as part of the initial SP05 introduction, not as a later add-on. Sources:
   - https://help.sap.com/doc/abapdocu_740_index_htm/7.40/en-US/abennews-740_sp05-mesh.htm
   - https://help.sap.com/doc/abapdocu_740_index_htm/7.40/en-US/abaptypes_mesh.htm

2. **Definitional coupling.** The 7.40 glossary defines a mesh as a structure "whose components ... are exclusively internal tables associated with each other using associations." A `BEGIN OF MESH ... END OF MESH` block with no `ASSOCIATION` field-line is semantically vacuous; the two cannot have been shipped in different releases.

3. **Horst Keller's SP05 announcement** (SCN, 2014-02-06, kernel 741) has a dedicated "Meshes" section describing them exactly as above. The prior SP02 news post has no mesh material; the later SP08 news series covers other features (FILTER/REDUCE/FOR, optional interface methods, etc.) with no mesh entry. Archive: https://web.archive.org/web/2015/http://scn.sap.com/community/abap/blog/2014/02/06/abap-news-for-release-740-sp05

4. **Negative evidence at 7.50 / 7.51.** Neither the 7.50 change overview (`abennews-750.htm`) nor the 7.51 change overview lists any mesh topic. The 7.50 sub-page "Data Types and ABAP Dictionary in Release 7.50" covers only int8, global temporary tables, and replacement objects — no mesh entry. There is no documented 7.51 enhancement to meshes that a narrower gate could correspond to.

5. **Internal inconsistency in abaplint.** In `type_mesh_begin.ts` and `type_mesh_end.ts` the guard is `Release.v740sp05`; only `type_mesh.ts` sets `Release.v751`. Under a `v750` target the block openers are accepted but no legal content is — which means no `MESH` block can parse at 7.50 at all, even though `BEGIN OF MESH` / `END OF MESH` are accepted at 7.40 SP05.

## Source pointer

`packages/core/src/abap/2_statements/statements/type_mesh.ts`:

```ts
const on = seq("ON", NamespaceSimpleName, "=", NamespaceSimpleName,
               star(seq("AND", NamespaceSimpleName, "=", NamespaceSimpleName)));
const using = seq("USING KEY", NamespaceSimpleName);
const association = seq("ASSOCIATION", NamespaceSimpleName, "TO", NamespaceSimpleName, plus(on));
const ret = ver(Release.v751,
                seq("TYPES", NamespaceSimpleName, "TYPE", opt("REF TO"), TypeName,
                    plus(association), opt(using)));
```

For comparison, `type_mesh_begin.ts`:

```ts
const ret = ver(Release.v740sp05, seq("TYPES", "BEGIN OF MESH", NamespaceSimpleName));
```

and `type_mesh_end.ts`:

```ts
const ret = ver(Release.v740sp05, seq("TYPES", "END OF MESH", NamespaceSimpleName));
```

None of the three files contains a comment explaining the choice of release. `type_mesh.ts` was introduced in PR #2266 ("2.82.6, bugfixes", merged 2021-11-30), whose body notes only "start adding MESH statement".

## Suggested fix

Change the guard in `packages/core/src/abap/2_statements/statements/type_mesh.ts` from `Release.v751` to `Release.v740sp05` so it aligns with the surrounding `BEGIN OF MESH` / `END OF MESH` statements and with the SAP-documented introduction release:

```ts
const ret = ver(Release.v740sp05,
                seq("TYPES", NamespaceSimpleName, "TYPE", opt("REF TO"), TypeName,
                    plus(association), opt(using)));
```

### Caveat about `USING KEY`

The one sub-clause I did not independently verify to 7.40 SP05 is `opt(using)` (i.e. `USING KEY <name>`). The 7.40 SP05 `abaptypes_mesh_association.htm` page appears to document `USING KEY` as part of the same feature set, but I did not confirm this against an independent 7.40 source. If it turns out that `USING KEY` specifically is a later addition, that could be split out with its own version guard while leaving the base `ASSOCIATION ... TO ... ON [AND ...]` form at `v740sp05`. Happy to follow up with more evidence if the maintainers want that distinction pinned down before merging.

## Prior art in the tracker

I searched the abaplint tracker for `"MESH ASSOCIATION"`, `type_mesh`, `parser_error mesh`, `v751 mesh`, and `mesh` — the only two hits containing `mesh` are issue #2970 (unrelated feature request, closed) and PR #2266 (initial mesh support, merged). No prior issue reports this specific version-gate discrepancy.

## Cross-references

- SAP 7.40 SP05 mesh introduction: https://help.sap.com/doc/abapdocu_740_index_htm/7.40/en-US/abennews-740_sp05-mesh.htm
- SAP 7.40 `TYPES - BEGIN OF MESH`: https://help.sap.com/doc/abapdocu_740_index_htm/7.40/en-US/abaptypes_mesh.htm
- SAP 7.50 `TYPES - BEGIN OF MESH` (unchanged): https://help.sap.com/doc/abapdocu_750_index_htm/7.50/en-US/abaptypes_mesh.htm
- abaplint `type_mesh.ts` (main): https://github.com/abaplint/abaplint/blob/main/packages/core/src/abap/2_statements/statements/type_mesh.ts
- abaplint `type_mesh_begin.ts` (main): https://github.com/abaplint/abaplint/blob/main/packages/core/src/abap/2_statements/statements/type_mesh_begin.ts
- abaplint `type_mesh_end.ts` (main): https://github.com/abaplint/abaplint/blob/main/packages/core/src/abap/2_statements/statements/type_mesh_end.ts

**Comment by larshp** [comment]
@rubber-duck-53 thanks, can you help adding your name and affiliation in your github profile? It feels a bit better to know who people is when spending my spare time maintaining the project

**Comment by rubber-duck-53** [comment]
@larshp profile updated. Fair enough to spend my time as well.

**Comment by larshp** [comment]
published in v2.119.63
